compiler/rustc_expand/src/module.rs

```diff
@@ -103,25 +103,47 @@ crate fn push_directory(
     id: Ident,
     attrs: &[Attribute],
     Directory { mut ownership, mut path }: Directory,
+    inline: Inline,
+    inner_span: Span,
+    span: Span, // The span to blame on errors.
 ) -> Directory {
-    if let Some(filename) = sess.first_attr_value_str_by_name(attrs, sym::path) {
-        path.push(&*filename.as_str());
-        ownership = DirectoryOwnership::Owned { relative: None };
-    } else {
-        // We have to push on the current module name in the case of relative
-        // paths in order to ensure that any additional module paths from inline
-        // `mod x { ... }` come after the relative extension.
-        //
-        // For example, a `mod z { ... }` inside `x/y.rs` should set the current
-        // directory path to `/x/y/z`, not `/x/z` with a relative offset of `y`.
-        if let DirectoryOwnership::Owned { relative } = &mut ownership {
-            if let Some(ident) = relative.take() {
-                // Remove the relative offset.
-                path.push(&*ident.as_str());
+    match inline {
+        Inline::Yes => {
+            if let Some(filename) = sess.first_attr_value_str_by_name(attrs, sym::path) {
+                path.push(&*filename.as_str());
+                ownership = DirectoryOwnership::Owned { relative: None };
+            } else {
+                // We have to push on the current module name in the case of relative
+                // paths in order to ensure that any additional module paths from inline
+                // `mod x { ... }` come after the relative extension.
+                //
+                // For example, a `mod z { ... }` inside `x/y.rs` should set the current
+                // directory path to `/x/y/z`, not `/x/z` with a relative offset of `y`.
+                if let DirectoryOwnership::Owned { relative } = &mut ownership {
+                    if let Some(ident) = relative.take() {
+                        // Remove the relative offset.
+                        path.push(&*ident.as_str());
+                    }
+                }
+                path.push(&*id.as_str());
+            }
+        }
+        Inline::No => {
+            // FIXME: This is a subset of `parse_external_mod` without actual parsing,
+            // check whether the logic for unloaded, loaded and inline modules can be unified.
+            if let Ok(mp) = submod_path(sess, id, span, &attrs, ownership, &path) {
+                ownership = mp.ownership;
             }
+
+            // Extract the directory path for submodules of  the module.
+            path = match sess.source_map().span_to_unmapped_path(inner_span) {
```

Comment:
Could this cause problems if a proc-macro expands to an outline module (e.g. `quote! { mod my_mod; }`)? I assume that we would want the path to be based on the actual on-disk directory structure, not the filename of the proc-macro that emitted the tokens.

Reply:
>Could this cause problems if a proc-macro expands to an outline module (e.g. `quote! { mod my_mod; }`)?

Probably not.
Inner span of the module will still be set by the compiler during module loading (to the file from which the module was actually loaded) and not by a proc macro.

#82415 (https://github.com/rust-lang/rust/pull/82415/commits/af40557491ae95e98fc249c88566ef990425573d specifically) still removes this logic though and operates on paths directly.